prune/prunables.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prune
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Some default pruneable functions
+
+// DefaultPodIsPrunable is a default IsPrunableFunc to be used specifically with Pod resources.
+// This can be overridden by registering your own IsPrunableFunc via the RegisterIsPrunableFunc method
+func DefaultPodIsPrunable(obj client.Object) error {
+	gvk := schema.GroupVersionKind{
+		Group:   "core",
+		Version: "v1",
+		Kind:    "Pod",
+	}
+
+	// if the object is not a Pod then it is not pruneable
+	if obj.GetObjectKind().GroupVersionKind() != gvk {
+		return &ErrUnprunable{
+			Obj:    &obj,
+			Reason: "Object is not a Pod",
+		}
+	}
```

Comment:
You're checking the Kubernetes type, but, while related, we care about the go type here. So you'd want something like:

```go
pod, ok := obj.(*corev1.Pod)
if !ok {
    return // a cool error
}
```

To add more detail:

```go 
var pod corev1.Pod
gvk := pod.GetObjectKind().GroupVersionKind()
```

After running that snippet, `gvk` will be an empty `GroupVersionKind` struct since its not set in the zero value Pod struct. This probably wouldn't happen in real life since you'll likely be reading the Pod from the server, which will return it with correct GVK information, but it demonstrates when the Kubernetes type and actual go type can drift.

One more thing, this is one of the rare situations that I think a `panic` or `log.Fatal` would be a better design than returning an error. Usually `panic`-ing in go is a bad practice, but I think this fits the narrow slice of usecases where it makes sense. Typically a `panic` means "something has happened that the program cannot recover from." If an operator-author incorrectly registers a pruneable function for the wrong GVK, that would be something that the program cannot recover from since it cannot correct a mistake the user made when configuring it. It also wouldn't make sense in my opinion to expose an error since someone consuming the operator is unable to fix that error since the code itself is broken and needs to be fixed. I would keep it simple and just use this:

```go
pod := obj.(*corev1.Pod)
```

A type-assertion will panic if it is not possible and there is only one return value. I think in this case, we are fine letting the program panic if the pod isPruneable func gets a non-pod object.

Reply:
Thanks for the detailed explanation! This makes more logical sense to me after reading your explanation so I will go ahead and work on implementing this for both the `Pod` and `Job` default prunable functions.